applications/serial_lte_modem/src/gnss/slm_at_gnss.c

```diff
@@ -0,0 +1,709 @@
+
+#if defined(CONFIG_SLM_PGPS)
+	int err2;
+
+	err2 = nrf_cloud_pgps_request_all();
+	if (err2) {
+		LOG_ERR("Failed to request P-GPS data: %d", err2);
+	}
+	err2 = nrf_cloud_pgps_notify_prediction();
+	if (err2) {
+		LOG_ERR("Failed to request notification of prediction: %d", err2);
+	}
+#endif /* CONFIG_SLM_PGPS */
+}
+
+/* NOTE this event handler runs in interrupt context */
+static void gnss_event_handler(int event)
+{
+	switch (event) {
+	case NRF_MODEM_GNSS_EVT_PVT:
+		LOG_DBG("NRF_MODEM_GNSS_EVT_PVT");
+		on_gnss_evt_pvt();
+		break;
+	case NRF_MODEM_GNSS_EVT_FIX:
+		LOG_INF("NRF_MODEM_GNSS_EVT_FIX");
+		on_gnss_evt_fix();
+		break;
+	case NRF_MODEM_GNSS_EVT_NMEA:
+		LOG_DBG("NRF_MODEM_GNSS_EVT_NMEA");
+		break;
+	case NRF_MODEM_GNSS_EVT_AGPS_REQ:
+		LOG_INF("NRF_MODEM_GNSS_EVT_AGPS_REQ");
+		on_gnss_evt_agps_req();
+		break;
+	case NRF_MODEM_GNSS_EVT_BLOCKED:
+		LOG_INF("NRF_MODEM_GNSS_EVT_BLOCKED");
+		break;
+	case NRF_MODEM_GNSS_EVT_UNBLOCKED:
+		LOG_INF("NRF_MODEM_GNSS_EVT_UNBLOCKED");
+		break;
+	case NRF_MODEM_GNSS_EVT_PERIODIC_WAKEUP:
+		LOG_DBG("NRF_MODEM_GNSS_EVT_PERIODIC_WAKEUP");
+		break;
+	case NRF_MODEM_GNSS_EVT_SLEEP_AFTER_TIMEOUT:
+		LOG_DBG("NRF_MODEM_GNSS_EVT_SLEEP_AFTER_TIMEOUT");
+		break;
+	case NRF_MODEM_GNSS_EVT_SLEEP_AFTER_FIX:
+		LOG_DBG("NRF_MODEM_GNSS_EVT_SLEEP_AFTER_FIX");
+		break;
+	case NRF_MODEM_GNSS_EVT_REF_ALT_EXPIRED:
+		LOG_DBG("NRF_MODEM_GNSS_EVT_REF_ALT_EXPIRED");
+		break;
+	default:
+		break;
+	}
+}
+
+#if defined(NRF_CLOUD_ACCESS)
+static void on_cloud_evt_ready(void)
+{
+	int err;
+	struct cloud_msg msg = {
+		.qos = CLOUD_QOS_AT_MOST_ONCE,
+		.endpoint.type = CLOUD_EP_STATE,
+		.buf = SERVICE_INFO_GPS,
+		.len = strlen(SERVICE_INFO_GPS)
+	};
+
+	/* Update nRF Cloud with GPS service info signifying GPS capabilities. */
+	err = cloud_send(nrf_cloud, &msg);
+	if (err) {
+		LOG_WRN("Failed to send message to cloud, error: %d", err);
+	}
+#if defined(CONFIG_SLM_PGPS)
+	/* Initialize P-GPS. */
+	struct nrf_cloud_pgps_init_param param = {
+		.event_handler = pgps_event_handler,
+		.storage_base = PM_MCUBOOT_SECONDARY_ADDRESS,
+		.storage_size = PM_MCUBOOT_SECONDARY_SIZE};
+
+	err = nrf_cloud_pgps_init(&param);
+	if (err) {
+		LOG_ERR("Error from P-GPS init: %d", err);
+	}
+#endif /* CONFIG_SLM_PGPS */
+
+#if defined(CONFIG_SLM_AGPS) || defined(CONFIG_SLM_PGPS)
+	/* Start GPS module, which would request back of A-GPS data. */
+	err = nrf_modem_gnss_start();
+	if (err) {
+		LOG_ERR("Failed to start GNSS, error: %d", err);
+	} else {
+		ttft_start = k_uptime_get();
+	}
+#endif /* CONFIG_SLM_AGPS || CONFIG_SLM_PGPS */
+
+#if defined(CONFIG_SLM_CELL_POS)
+	k_work_submit(&cell_pos_req);
+#endif /* CONFIG_SLM_CELL_POS */
+}
+
+static void on_cloud_evt_data_received(const struct cloud_event *const evt)
+{
+#if defined(CONFIG_SLM_AGPS)
+		int err1;
+
+		err1 = nrf_cloud_agps_process(evt->data.msg.buf, evt->data.msg.len, NULL);
+		if (err1) {
+			LOG_INF("Unable to process A-GPS data, error: %d", err1);
+		}
```

Comment:
If there is no error, then this should just return, and not continue on to pass the same data to PGPS and CELL_POS, if either or both are also enabled.  Not sure if enabling CELL_POS plus AGPS or PGPS makes any sense, but certainly AGPS + PGPS is possible.

Same for the PGPS below.

Reply:
Updated with no code logic that would prevent multiple handling of one nRF Cloud data.

Reply:
This is better, but does not allow for AGPS + PGPS to both be enabled in the Kconfig settings and function together properly.  Maybe adding this is unnecessary for this application?

When combined, AGPS data will be used for the smaller assistance items, if the modem needs them and if there is a cloud connection.  It uses the PGPS data for the large ephemerides assistance.  If the cloud connection is not available when trying to get a fix, the PGPS library will use the current system clock for time assistance and the last saved fix location for rough location assistance.

If you want to add this capability, you should add a `RUN_TYPE_AGPS_PGPS`, and handle that case.  When data is received from the cloud, we do not know what the format is, so we first give it to the AGPS processor, and if it doesn't like it, give it to the PGPS processor:
```
} else if (run_type == RUN_TYPE_AGPS_PGPS) {
        err = nrf_cloud_agps_process(evt->data.msg.buf, evt->data.msg.len, NULL);
        if (err) { /* not the right format for AGPS; try passing to PGPS instead */
		err = nrf_cloud_pgps_process(evt->data.msg.buf, evt->data.msg.len);
		if (err) {
			LOG_ERR("Unable to process P-GPS data, error: %d", err);
		}
        }